Add progress bar web component

andresriveratoro opened the pull request:
## Description
Closes https://github.com/department-of-veterans-affairs/va.gov-team/issues/32866

## Testing done


## Screenshots


## Acceptance criteria
- [ ]

## Definition of done
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

pdavies88 reviewed (approved):
Looks great to me when testing it

k80bowman reviewed (approved):
This seems good to me as well. @bkjohnson any comments?

bkjohnson commented on packages/web-components/src/components/va-progress-bar/test/va-progress-bar.e2e.ts:
```diff
@@ -0,0 +1,75 @@
+    await axeCheck(page);
+  });
+
+  it('fires an analytics event when enable-analytics is set and percent is set to 0', async () => {
```
It might be nice if these tests included the event details as a way of documenting what is included.

bkjohnson reviewed (approved):
Looks good! Only one minor comment about testing.

andresriveratoro commented:
Updated the aria label default value to use percent if a label isn't provided and updated CSS to prevent inner bar from growing over 100% if given a percent value over 100.

bkjohnson commented on packages/web-components/src/components/va-progress-bar/va-progress-bar.tsx:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  render() {
+    const ariaLabel = this.label ?? `${this.percent}%`;
```
Looks like this is causing a test to fail.

Also, instead of nullish coalescing, when we're going for default values can we set that on the property itself, or use destructuring with default assignment?

andresriveratoro replied:
I've pushed up a fix for the tests. I think the default value that we want for label is the value of the percent prop, but we expose a label prop to allow a developer to override it.

andresriveratoro replied:
How about this?

```
  componentWillLoad() {
    if (!this.label) this.label = `${this.percent}`;
  }
```

andresriveratoro replied:
I'm against the previous componentWillLoad idea after implementing it locally because it's adding unnecessary complexity.

bkjohnson replied:
What about something like this, but "complete" instead of "Working"? This diff is a bit awkward because I moved `percent` above `label` :upside_down_face: 

![image](https://user-images.githubusercontent.com/2008881/145312608-6855847c-cdce-4790-805f-088a90b74e06.png)

andresriveratoro replied:
It looks like this approach doesn't work.
<img width="718" alt="Screen Shot 2021-12-09 at 09 47 43" src="https://user-images.githubusercontent.com/36863582/145418182-e70d90ec-4c56-4846-b922-ee24ac476351.png">

andresriveratoro replied:
I've updated it to use destructuring with default assignment.

andresriveratoro commented:
One of the tests was updated to skip until we discover how to properly test lifecycle methods on component initialization.

bkjohnson commented on packages/web-components/src/components/va-progress-bar/va-progress-bar.tsx:
```diff
@@ -12,14 +20,19 @@ export class VaProgressBar {
+   * The text label for the progress bar.
    */
-  @Prop() percent: number;
+  @Prop({ mutable: true }) label: string = `${this.percent}% complete`;
+
+  @Watch('percent')
+  updateLabel(newPercent: number) {
+    this.label = `${newPercent}% complete`;
+  }
```
Having a mutable prop with  `@Watch` looks like it might override a custom `label` value set by an engineer.

andresriveratoro replied:
You're correct. This isn't going to work.

bkjohnson commented on packages/web-components/src/components/va-progress-bar/va-progress-bar.tsx:
```diff
@@ -34,15 +47,15 @@ export class VaProgressBar {
         action: 'change',
         details: {
-          label: this.label || `${this.percent}% complete`,
+          label: this.label,
```
Is the `@Watch` decorator what's making sure that this label is defined?

andresriveratoro replied:
I added the watch decorator to make sure the analytics receives the correct details if/when the percent prop is updated, but like you mentioned above, this won't work if a developer sets their own label because it'll override it.
